README.md
# KISS FFT

KISS FFT - A mixed-radix Fast Fourier Transform based up on the principle, 
"Keep It Simple, Stupid."

There are many great fft libraries already around. Kiss FFT is not trying
to be better than any of them. It only attempts to be a reasonably efficient, 
moderately useful FFT that can use fixed or floating data types and can be 
incorporated into someone's C program in a few minutes with trivial licensing.

## Basic usage

    The basic usage for 1-d complex FFT is:

```
        #include "kiss_fft.h"

        kiss_fft_cfg cfg = kiss_fft_alloc( nfft, is_inverse_fft, 0, 0 );

            // put kth sample in cx_in[k].r and cx_in[k].i
            
            kiss_fft( cfg , cx_in , cx_out );
            
            // transformed. DC is in cx_out[0].r and cx_out[0].i 
            
        free(cfg);
```

Note: frequency-domain data is stored from dc up to 2pi.
So cx_out[0] is the dc bin of the FFT and cx_out[nfft/2]
is the Nyquist bin (if exists).

Declarations are in "kiss_fft.h", along with a brief description of the 
functions you'll need to use.

Code definitions for 1d complex FFTs are in kiss_fft.c.

## Real valued FFT

A real valued optimized FFT which takes real valued signals as its input is implemtned in `kiss_fftr.h` and `kiss_fftr.c`. It returns the positive half-spectrum: (nfft/2+1) complex frequency bins.

## Unit tests
Run `make testall` which compares the kiss with fftw3 and does
testing in python.

## Advanced topics

You can do other cool stuff with the extras you'll find in `tools/`

* multi-dimensional FFTs 
* fast convolution (not available for fixed point)
* spectrum image creation

The core fft and most `tools/` code can be compiled using float, double,
 Q15 short or Q31 samples. The default is float.


## Attribution
This is a fork from the original kiss-fft library by Mark Borgerding
<Mark@Borgerding.net>.


CHANGELOG
1.4.0 22 Dec 2017, mail@berndporr.me.uk
  Removed the SIMD stuff as it's really not very Kiss-like
  Removed the C++ stuff

1.3.0 2012-07-18
  removed non-standard malloc.h from kiss_fft.h

  moved -lm to end of link line

  checked various return values

  converted python Numeric code to NumPy
 
  fixed test of int32_t on 64 bit OS

  added padding in a couple of places to allow SIMD alignment of structs

1.2.9 2010-05-27
  threadsafe ( including OpenMP )

  first edition of kissfft.hh the C++ template fft engine
 
1.2.8 
  Changed memory.h to string.h -- apparently more standard
  
  Added openmp extensions.  This can have fairly linear speedups for larger FFT sizes.

1.2.7 
  Shrank the real-fft memory footprint. Thanks to Galen Seitz.

1.2.6 (Nov 14, 2006) The "thanks to GenArts" release.
  Added multi-dimensional real-optimized FFT, see tools/kiss_fftndr
  Thanks go to GenArts, Inc. for sponsoring the development.

1.2.5 (June 27, 2006) The "release for no good reason" release.
   Changed some harmless code to make some compilers' warnings go away.
   Added some more digits to pi -- why not.
   Added kiss_fft_next_fast_size() function to help people decide how much to pad.
   Changed multidimensional test from 8 dimensions to only 3 to avoid testing 
   problems with fixed point (sorry Buckaroo Banzai).

1.2.4 (Oct 27, 2005)   The "oops, inverse fixed point real fft was borked" release. 
   Fixed scaling bug for inverse fixed point real fft -- also fixed test code that should've been failing.
    Thanks to Jean-Marc Valin for bug report.

   Use sys/types.h for more portable types than short,int,long => int16_t,int32_t,int64_t
   If your system does not have these, you may need to define them -- but at least it breaks in a 
   loud and easily fixable way -- unlike silently using the wrong size type.

   Hopefully tools/psdpng.c is fixed -- thanks to Steve Kellog for pointing out the weirdness.

1.2.3 (June 25, 2005)   The "you want to use WHAT as a sample" release.
    Added ability to use 32 bit fixed point samples -- requires a 64 bit intermediate result, a la 'long long'

    Added ability to do 4 FFTs in parallel by using SSE SIMD instructions. This is accomplished by
    using the __m128 (vector of 4 floats) as kiss_fft_scalar.  Define USE_SIMD to use this.
    
    I know, I know ...  this is drifting a bit from the "kiss" principle, but the speed advantages 
    make it worth it for some.  Also recent gcc makes it SOO easy to use vectors of 4 floats like a POD type.

1.2.2 (May 6, 2005)   The Matthew release
    Replaced fixed point division with multiply&shift.  Thanks to Jean-Marc Valin for 
    discussions regarding.  Considerable speedup for fixed-point.

    Corrected overflow protection in real fft routines  when using fixed point.
    Finder's Credit goes to Robert Oschler of robodance for pointing me at the bug.
    This also led to the CHECK_OVERFLOW_OP macro.

1.2.1 (April 4, 2004) 
    compiles cleanly with just about every -W warning flag under the sun

    reorganized kiss_fft_state so it could be read-only/const. This may be useful for embedded systems
    that are willing to predeclare twiddle factors, factorization.

    Fixed C_MUL,S_MUL on 16-bit platforms.

    tmpbuf will only be allocated if input & output buffers are same
    scratchbuf will only be allocated for ffts that are not multiples of 2,3,5
 
    NOTE: The tmpbuf,scratchbuf changes may require synchronization code for multi-threaded apps.


1.2 (Feb 23, 2004)
    interface change -- cfg object is forward declaration of struct instead of void*
    This maintains type saftey and lets the compiler warn/error about stupid mistakes.
            (prompted by suggestion from Erik de Castro Lopo)

    small speed improvements

    added psdpng.c -- sample utility that will create png spectrum "waterfalls" from an input file
        ( not terribly useful yet)

1.1.1 (Feb 1, 2004 )
    minor bug fix -- only affects odd rank, in-place, multi-dimensional FFTs

1.1 : (Jan 30,2004)
    split sample_code/ into test/ and tools/

    Removed 2-D fft and added N-D fft (arbitrary)

    modified fftutil.c to allow multi-d FFTs

    Modified core fft routine to allow an input stride via kiss_fft_stride()
    (eased support of multi-D ffts)

    Added fast convolution filtering (FIR filtering using overlap-scrap method, with tail scrap)

    Add kfc.[ch]: the KISS FFT Cache. It takes care of allocs for you ( suggested by Oscar Lesta ).

1.0.1 (Dec 15, 2003)
    fixed bug that occurred when nfft==1. Thanks to Steven Johnson.
    
1.0 : (Dec 14, 2003)
    changed kiss_fft function from using a single buffer, to two buffers.
    If the same buffer pointer is supplied for both in and out, kiss will
    manage the buffer copies.

    added kiss_fft2d and kiss_fftr as separate source files (declarations in kiss_fft.h )

0.4 :(Nov 4,2003) optimized for radix 2,3,4,5

0.3 :(Oct 28, 2003) woops, version 2 didn't actually factor out any radices other than 2.
        Thanks to Steven Johnson for finding this one.

0.2 :(Oct 27, 2003) added mixed radix, only radix 2,4 optimized versions

0.1 :(May 19 2003)  initial release, radix 2 only


_kiss_fft_guts.h
/*
Copyright (c) 2003-2010, Mark Borgerding

All rights reserved.

Redistribution and use in source and binary forms, with or without modification, are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright notice, this list of conditions and the following disclaimer.
    * Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the following disclaimer in the documentation and/or other materials provided with the distribution.
    * Neither the author nor the names of any contributors may be used to endorse or promote products derived from this software without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/

/* kiss_fft.h
   defines kiss_fft_scalar as either short or a float type
   and defines
   typedef struct { kiss_fft_scalar r; kiss_fft_scalar i; }kiss_fft_cpx; */
#include "kiss_fft.h"
#include <limits.h>

#define MAXFACTORS 32
/* e.g. an fft of length 128 has 4 factors 
 as far as kissfft is concerned
 4*4*4*2
 */

struct kiss_fft_state{
    int nfft;
    int inverse;
    int factors[2*MAXFACTORS];
    kiss_fft_cpx twiddles[1];
};

/*
  Explanation of macros dealing with complex math:

   C_MUL(m,a,b)         : m = a*b
   C_FIXDIV( c , div )  : if a fixed point impl., c /= div. noop otherwise
   C_SUB( res, a,b)     : res = a - b
   C_SUBFROM( res , a)  : res -= a
   C_ADDTO( res , a)    : res += a
 * */
#ifdef FIXED_POINT
#if (FIXED_POINT==32)
# define FRACBITS 31
# define SAMPPROD int64_t
#define SAMP_MAX 2147483647
#else
# define FRACBITS 15
# define SAMPPROD int32_t 
#define SAMP_MAX 32767
#endif

#define SAMP_MIN -SAMP_MAX

#if defined(CHECK_OVERFLOW)
#  define CHECK_OVERFLOW_OP(a,op,b)  \
	if ( (SAMPPROD)(a) op (SAMPPROD)(b) > SAMP_MAX || (SAMPPROD)(a) op (SAMPPROD)(b) < SAMP_MIN ) { \
		fprintf(stderr,"WARNING:overflow @ " __FILE__ "(%d): (%d " #op" %d) = %ld\n",__LINE__,(a),(b),(SAMPPROD)(a) op (SAMPPROD)(b) );  }
#endif


#   define smul(a,b) ( (SAMPPROD)(a)*(b) )
#   define sround( x )  (kiss_fft_scalar)( ( (x) + (1<<(FRACBITS-1)) ) >> FRACBITS )

#   define S_MUL(a,b) sround( smul(a,b) )

#   define C_MUL(m,a,b) \
      do{ (m).r = sround( smul((a).r,(b).r) - smul((a).i,(b).i) ); \
          (m).i = sround( smul((a).r,(b).i) + smul((a).i,(b).r) ); }while(0)

#   define DIVSCALAR(x,k) \
	(x) = sround( smul(  x, SAMP_MAX/k ) )

#   define C_FIXDIV(c,div) \
	do {    DIVSCALAR( (c).r , div);  \
		DIVSCALAR( (c).i  , div); }while (0)

#   define C_MULBYSCALAR( c, s ) \
    do{ (c).r =  sround( smul( (c).r , s ) ) ;\
        (c).i =  sround( smul( (c).i , s ) ) ; }while(0)

#else  /* not FIXED_POINT*/

#   define S_MUL(a,b) ( (a)*(b) )
#define C_MUL(m,a,b) \
    do{ (m).r = (a).r*(b).r - (a).i*(b).i;\
        (m).i = (a).r*(b).i + (a).i*(b).r; }while(0)
#   define C_FIXDIV(c,div) /* NOOP */
#   define C_MULBYSCALAR( c, s ) \
    do{ (c).r *= (s);\
        (c).i *= (s); }while(0)
#endif

#ifndef CHECK_OVERFLOW_OP
#  define CHECK_OVERFLOW_OP(a,op,b) /* noop */
#endif

#define  C_ADD( res, a,b)\
    do { \
	    CHECK_OVERFLOW_OP((a).r,+,(b).r)\
	    CHECK_OVERFLOW_OP((a).i,+,(b).i)\
	    (res).r=(a).r+(b).r;  (res).i=(a).i+(b).i; \
    }while(0)
#define  C_SUB( res, a,b)\
    do { \
	    CHECK_OVERFLOW_OP((a).r,-,(b).r)\
	    CHECK_OVERFLOW_OP((a).i,-,(b).i)\
	    (res).r=(a).r-(b).r;  (res).i=(a).i-(b).i; \
    }while(0)
#define C_ADDTO( res , a)\
    do { \
	    CHECK_OVERFLOW_OP((res).r,+,(a).r)\
	    CHECK_OVERFLOW_OP((res).i,+,(a).i)\
	    (res).r += (a).r;  (res).i += (a).i;\
    }while(0)

#define C_SUBFROM( res , a)\
    do {\
	    CHECK_OVERFLOW_OP((res).r,-,(a).r)\
	    CHECK_OVERFLOW_OP((res).i,-,(a).i)\
	    (res).r -= (a).r;  (res).i -= (a).i; \
    }while(0)


#ifdef FIXED_POINT
#  define KISS_FFT_COS(phase)  floor(.5+SAMP_MAX * cos (phase))
#  define KISS_FFT_SIN(phase)  floor(.5+SAMP_MAX * sin (phase))
#  define HALF_OF(x) ((x)>>1)
#else
#  define KISS_FFT_COS(phase) (kiss_fft_scalar) cos(phase)
#  define KISS_FFT_SIN(phase) (kiss_fft_scalar) sin(phase)
#  define HALF_OF(x) ((x)*.5)
#endif

#define  kf_cexp(x,phase) \
	do{ \
		(x)->r = KISS_FFT_COS(phase);\
		(x)->i = KISS_FFT_SIN(phase);\
	}while(0)


/* a debugging function */
#define pcpx(c)\
    fprintf(stderr,"%g + %gi\n",(double)((c)->r),(double)((c)->i) )


#ifdef KISS_FFT_USE_ALLOCA
// define this to allow use of alloca instead of malloc for temporary buffers
// Temporary buffers are used in two case: 
// 1. FFT sizes that have "bad" factors. i.e. not 2,3 and 5
// 2. "in-place" FFTs.  Notice the quotes, since kissfft does not really do an in-place transform.
#include <alloca.h>
#define  KISS_FFT_TMP_ALLOC(nbytes) alloca(nbytes)
#define  KISS_FFT_TMP_FREE(ptr) 
#else
#define  KISS_FFT_TMP_ALLOC(nbytes) KISS_FFT_MALLOC(nbytes)
#define  KISS_FFT_TMP_FREE(ptr) KISS_FFT_FREE(ptr)
#endif
